File: Windows/FireflyDevice/FDString.cpp
//
//  FDString.cpp
//  FireflyDevice
//
//  Created by Denis Bohm on 3/27/14.
//  Copyright (c) 2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

#include "FDString.h"

#include <cstdarg>
#include <cstdio>
#include <iomanip>
#include <sstream>
#include <vector>

namespace FireflyDesign {

	std::string FDString::format(const std::string fmt, ...) {
		std::vector<char> str(100, '\0');
		while (1) {
			va_list ap;
			va_start(ap, fmt);
			auto n = vsnprintf_s(str.data(), str.size(), _TRUNCATE, fmt.c_str(), ap);
			va_end(ap);
			if ((n > -1) && (size_t(n) < str.size())) {
				return str.data();
			}
			if (n > -1)
				str.resize(n + 1);
			else
				str.resize(str.size() * 2);
		}
		return str.data();
	}

	std::string FDString::formatDateTime(time_type epochTime) {
		std::time_t time = (std::time_t)epochTime;
		struct tm tm;
		gmtime_s(&tm, &time);
		std::stringstream ss;
		// "yyyy-MM-dd HH:mm:ss.SSS"
		ss << std::put_time(&tm, "%Y-%m-%d %H:%M:%S");
		std::string s = ss.str();
		return s;
	}

}

File: Android/Libraries/fireflydevice/src/main/java/com/fireflydesign/fireflydevice/FDString.java
//
//  FDString.cpp
//  FireflyDevice
//
//  Created by Denis Bohm on 3/27/14.
//  Copyright (c) 2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

package com.fireflydesign.fireflydevice;

import java.nio.ByteBuffer;
import java.text.SimpleDateFormat;
import java.util.GregorianCalendar;

public class FDString {

    public static String format(String fmt, Object ... arguments) {
        if (arguments.length == 0) {
            return fmt;
        }
        try {
            return String.format(fmt, arguments);
        } catch (Exception e) {
            return fmt;
        }
    }

    public static String formatDateTime(double time) {
		// "yyyy-MM-dd HH:mm:ss.SSS"
        GregorianCalendar calendar = new GregorianCalendar();
        calendar.setTimeInMillis((long) (time * 1000.0));
        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
        return formatter.format(calendar.getTime());
	}

    static int parseInt(String string) {
        if (string.startsWith("0x") || string.startsWith("0X")) {
            return Integer.parseInt(string.substring(2), 16);
        }
        return Integer.parseInt(string);
    }

    // For example: 5f8e87b7ecf9558d8704e3af7177388098387368
    static byte[] parseBytes(String string) {
        int byteCount = string.length() / 2;
        ByteBuffer buffer = ByteBuffer.allocate(byteCount);
        for (int i = 0; i < string.length(); i += 2) {
            String substring = string.substring(i, i + 2);
            byte b = (byte)Integer.parseInt(substring, 16);
            buffer.put(b);
        }
        return buffer.array();
    }

}